### PR TITLE
Respect x robots tag by default

### PR DESCRIPTION
**Pull request by Stealcase**
Solves https://github.com/rom1504/img2dataset/issues/247 and https://github.com/rom1504/img2dataset/issues/245
and completes the previous PR: https://github.com/rom1504/img2dataset/pull/218

# Problem 
Currently the default behaviour of img2dataset is to ignore x-robots-tag directives.
It does not honor "nofollow" and "noindex", nor "noai" and "noimageai".

Many websites are adding clauses to their terms of service that using images marked with "noai" . Here's one from [Artstation](https://www.artstation.com/tos), owned by Epic Games:
image

Getting in trouble with EPIC Games because you used a popular Image downloading tool doesn't seem like a sensible default option. 

As mentioned, this is a quickly developing space, and since img2dataset is a very popular tool it would be beneficial if the commonly used tags were honoured. 

## Caveats
image
I don't think it's a good idea to remove the feature and completely "turn off" respecting directives. 

This will likely lead to more forks, and making development more inefficient. 

**Instead, we allow users to disable the default directives by passing in an empty list.** 
This, of course, involves legal risk, so should not be done lightly. But it ensures that people don't fork this repo just to get this simple feature. 

It also allows users to pass in additional filters if needed, if more directives are added in the future. 

### Other considerations
* This PR does NOT check "robots.txt", as it is out of scope, and should probably be implemented in a way that allows caching to improve performance, and avoid multiple calls to "/robots.txt" per website. 

* Testing was already performed by https://github.com/raincoastchris when he made this functionality, and this is manly a change in the defaults + Documentation.

### CODE REVIEW

File: img2dataset/main.py [hunk 1]
```diff
@@ -86,7 +86,7 @@ def download(
     incremental_mode: str = "incremental",
     max_shard_retry: int = 1,
     user_agent_token: Optional[str] = None,
-    disallowed_header_directives: Optional[List[str]] = None,
+    disallowed_header_directives: List[str] = ["noai", "noimageai", "noindex", "noimageindex"],
```
Comment (rom1504) [inline img2dataset/main.py:89]: Using a list as a default value is an anti pattern in python, put back None here
and instead do 
```
if disallowed_header_directives is None:
  disallowed_header_directives = ["noai", "noimageai", "noindex", "noimageindex"]
```
below

File: dataset_examples/laion-art-ignore-directives.md [hunk 2]
```diff
@@ -5,7 +5,9 @@ See [full description](https://github.com/LAION-AI/laion-datasets/blob/main/laio
 
 It is available at https://huggingface.co/datasets/laion/laion-art
 
-A good use case is to train an image generation model. However, concerns have been raised about how to ethically source training data at scale for such purposes, especially where creator consent may not have been expressed in advance of dataset construction. One solution is to allow artists to opt their images out of such usage by sending HTTP header directives when image content is downloaded. img2dataset can be configured to automatically respect such directives.
+Artists have raised ethical issues regarding the downloading of mass ammounts of image date for AI training without consent, so img2dataset respects emerging AI image header directives by default. However if you want to ignore these headers, you can pass in an empty list to the argument "dissalowed_header_directives".
+
```
Comment (rom1504) [inline dataset_examples/laion-art-ignore-directives.md:9]: fix typo, "dissalow"

> Turning this off can lead to potential legal risk

that is not true, let's remove this part

File: dataset_examples/laion-art-ignore-directives.md [hunk 3]
```diff
@@ -5,7 +5,9 @@ See [full description](https://github.com/LAION-AI/laion-datasets/blob/main/laio
 
 It is available at https://huggingface.co/datasets/laion/laion-art
 
-A good use case is to train an image generation model. However, concerns have been raised about how to ethically source training data at scale for such purposes, especially where creator consent may not have been expressed in advance of dataset construction. One solution is to allow artists to opt their images out of such usage by sending HTTP header directives when image content is downloaded. img2dataset can be configured to automatically respect such directives.
+Artists have raised ethical issues regarding the downloading of mass ammounts of image date for AI training without consent, so img2dataset respects emerging AI image header directives by default. However if you want to ignore these headers, you can pass in an empty list to the argument "dissalowed_header_directives".
```
Comment (rom1504) [inline dataset_examples/laion-art-ignore-directives.md:8]: fix typos

File: img2dataset/main.py (hunk 1 above)
Comment (Stealcase) [reply img2dataset/main.py]: Thanks for the tip 👍

File: dataset_examples/laion-art-ignore-directives.md [hunk 4]
```diff
@@ -5,7 +5,9 @@ See [full description](https://github.com/LAION-AI/laion-datasets/blob/main/laio
 
 It is available at https://huggingface.co/datasets/laion/laion-art
```
Comment (rom1504) [inline dataset_examples/laion-art-ignore-directives.md:6]: please add a few lines to explain how keeping disallowed_header_directives to the default of ["noai", "noimageai", "noindex", "noimageindex"] will help 1% of humans keep their privilege and exclude 99% of humans the ability to create art.

Comment (Stealcase) [reply dataset_examples/laion-art-ignore-directives.md:6]: This seems a little snarky. 

If you want to draw anything right now, you have created art. This won't change that. 

I don't understand how this changes how many people can make art? This won't gatekeep the people who want to ignore artists anyway: they can just disable the feature.

File: dataset_examples/laion-art-ignore-directives.md (hunk 3 above)
Comment (Stealcase) [reply dataset_examples/laion-art-ignore-directives.md]: Fixed 👍

**Comment by Stealcase** [comment]
Appreciate the speed you reviewed the PR. 😃 👍

File: dataset_examples/laion-art-ignore-directives.md (hunk 4 above)
Comment (rom1504) [reply dataset_examples/laion-art-ignore-directives.md:6]: Trying to slow down the creation of image generation image may slow down access of these models to people.

Most people do not have a talent for art. Such generative models allow them to create art.

So yes, removing access to artistic picture from model training sets will help artists keep their privilege and prevent most people from creating art.

Comment (Stealcase) [reply dataset_examples/laion-art-ignore-directives.md:6]: Okay, but most people also do not have the talent for playing football at a professional level. 

Are we preventing them from playing football? No, you can play football. You just aren't good enough to play on the National Team. 
You can play football or create art without being a professional, nobody is stopping you. It's one of the cheapest hobbies out there. 

Anybody can create art. It's not a new invention of 2022. Most artists who work professionally have been drawing for 10+ years to become skilled, it's not a talent they are born with. 

If you're saying that these image generation models are only viable and can only create art when using images by artists, are you not inherently saying that they need to art to work?

Again, this PR is only about the DEFAULT. You can (and probably will) ignore it.

**Comment by AUSBird** [comment]
I would love to see this PR merged! Great work!

I’m seeing a little bit of push back from the repo owner, I don’t see why there is an issue with accepting some people don’t want to partake in training other’s AI with their hard work.

I would love to hear your view though :)

**Comment by rom1504** [comment]
I clarified the readme and fixed the code. It is now the default to not download opted out images.

My opinion is that it is much more ethical to train AI model on as much content as possible to let more people benefit from the democratization of skills and art.

However, let's try and respect the opt out choice of people.

**Comment by alexjc** [comment]
Hi @rom1504,

First, I'd like to thank you for handling this PR despite your objections.  It wasn't graceful, but it got done!

Second, I want to share some unsolicited advice.  I left some time in between the PR and posting so it's not unnecessary drama.


I'd like to caution you for certain words and behaviors.  In this thread, it appears that:
- You personally value training "AI model on as much content as possible" over respecting opt-out.
- You implicitly encourage your users to break Terms of Service (e.g. of ArtStation).
- You incorrectly state there is no legal basis for opt-out, when it's mandated in Directive (EU) 2019/790 among others.
- You further imply that your users could/should violate these regulations to obtain more data.

Since you've integrated this PR despite personal objections, it's safe to say you understand the legal need for it.  Please also understand that a project owner's intentions and words can also get them into trouble legally.  Such words could be used as evidence against your organization(s) (e.g. LAION) and also make you personally liable for damages caused by your tool to third-parties downstream (i.e. organisations not bound by the project license).


Third, please consider that (i) "the ends justify the means" and (ii) one should violate the consent of others in order to pursue one's own personal goals are a Mind Virus based on logical fallacies.  The people who have adopted this ideology, especially those in the social hierarchy above you, are much more careful with their words.  Some of them are psychopaths, and will throw you under the bus legally if it's beneficial for them.

You should consider either rethinking your perspective (incl. with help) or learning to lie more consistently so you don't get in trouble!

All the best.

**Comment by rom1504** [comment]
> it's safe to say you understand the legal need for it.

I would be curious for you to present a convincing argument for this. It seems highly unclear whether anyone can legally restrict some data on the internet to be accessible only to human eyes.

It is however pretty clear that slowing down open AI is unethical.

Do you wish to encourage the default path of no open sourcing happening and all organizations keeping models and datasets as private as possible?
That will definitely happen if enough people like you make sure that open models and open data becomes illegal.

**Comment by AUSBird** [comment]
Thing is those images being used aren't open. So it’s unethical to make the AI using said unlicensed material.

But for the sake of this, I will for a second *pretend* that stopping AI from learning is unethical. It’s really not, but I will humour you.
The point is you are still being highly unethical about how you train that AI, cause you are taking the hard work of millions of artists, and trampling all over them by building something to do what they spent years perfecting while yelling that those who resist you trampling over them are “being unethical for stopping AI’s progress”

Fact is without these artists, this tool can’t exist period. You really should license their work, but at the very least let them opt out of this.
I see a lawsuit going through court that will decide if AI can train on artists’s work, and honestly, if we can’t agree of some way to help make everyone happy (You get as much data as you can, while letting artists say “No thanks, I want out”) then in all honesty I hope the artists win that suit.

**Comment by rom1504** [comment]
So indeed you would like the world to have only models that are not released and datasets that are not released.

Well you're in for a treat then, that's going to happen quite likely.
OpenAI will put dalle3 under an API and you won't know which pictures they used.

**Comment by rom1504** [comment]
There is no lawsuit against companies using private datasets and models

**Comment by AUSBird** [comment]
https://stablediffusionlitigation.com/
If this works out or not is to be seen, but I hope it succeeds.
Regulate the images you use or be regulated. This is my point.

**Comment by rom1504** [comment]
You can't argue in isolation from the real world. It doesn't work.

Choices of a few artists to fight against open datasets and open models are doing exactly that. The fact their claim is actually about opt out vs not opt out doesn't make a big difference.

The end result is organization will publish less to avoid any potential risk.

The opt out vs not opt out topic will then obviously be impossible to solve since the data will be fully privately collected

**Comment by rom1504** [comment]
https://stablediffusionlitigation.com/ does not concern any of the companies doing private datasets and models

**Comment by AUSBird** [comment]
I would love to see private models also hit because they are essentially in the same boat just hidden by shadows. Problem is the light won’t easily be put on how that sausage is made. 
Regulation on private models should exist for more reasons than artists wishes. I personally want none of my data or data about me in them. (Call me selfish)

My point is that you chose to make a public dataset, and I am not saying that’s a bad thing. I think it’s good to make a public dataset. But you also publicly made a fuss about ignoring the wishes of copyright holders.

**Comment by rom1504** [comment]
I still disagree fundamentally on the fact that content present on the internet should be consumed only by human eyes.

But you're missing the other point I was making. The action of you and people agreeing with you on attacking open data and open models does not happen in isolation.

Attacks on ML and especially open ML will result in a slow down first on open ML second on ML 
 which is a strict negative for anyone involved.

Art station and similar website solution is to say "we ban everything by default". It's not a constructive solution. To satisfy the complaints of a few they are making everyone result worse.

It's a good thing the rest of the internet is, as of now, not as misguided and has not decided to ban collection of data.

**Comment by AUSBird** [comment]
> Art station and similar website solution is to say "we ban everything by default". It's not a constructive solution.
I wholeheartedly agree. Without a doubt.
I also agree that people on twitter (Pro or Anti AI alike) screaming isn’t constructive.
That’s why I’m debating here with you. Cause it’s more constructive talking to someone who contributes to the thing in question. 

> I still disagree fundamentally on the fact that content present on the internet should be consumed only by human eyes.
That’s fine, but you too seem to be either missing or ignoring the fact we are talking about copyright work. Yes a human sees it and learns by looking from said creative work.
Yes creative work can be highly derivative.
However I disagree that training AI on the creative and copyright protected work without either a license or permission is more ethical than filtering it out.

Why should collecting images and making a dataset of copyrighted works be allowed when I can’t make a website or something to display that work? Why should your field be above copyright?

Otherwise I do agree the hit to open models and datasets is likely not the best, but it’s been made clear that these datasets and models don’t care what as you call “the few artists” want.

**Comment by AUSBird** [comment]
On that, it's not really an attack on all of AI. It's an attack on this particular model that uses data that was unlicensed or without permission.
Other open models can exist without being shot down by using data that either doesn't need a license, using it with a license or by asking for permission. 

Yes it does not stop private models and datasets. And that's unfortunate.

**Comment by rom1504** [comment]
> Why should collecting images and making a dataset of copyrighted works be allowed when I can’t make a website or something to display that work? Why should your field be above copyright?

You absolutely can collect website and display them. See https://www.google.com/webhp?output=search&tbm=isch&tbo=u&sa=X 
and  https://www.bing.com/images/feed

**Comment by rom1504** [comment]
> Other open models can exist without being shot down by using data that either doesn't need a license, using it with a license or by asking for permission.

Are you sure ? Can you point to one such model ?

**Comment by AUSBird** [comment]
> Are you sure ? Can you point to one such model ?

*facepalms*

**Pull request merged**